model/model_training/models/patching_utils.py

```diff
@@ -7,11 +7,14 @@ def compute_flash_attention(flash_attn, q, k, v, attention_mask=None, head_mask=
     # attention_mask (float): [bs, seq_len]
     batch_size, max_len = q.size(0), q.size(1)
 
-    qkv = torch.stack([q, k, v], dim=2).to(torch.float16)  # need to truncate in case input is fp32
+    qkv = torch.stack([q, k, v], dim=2)
+    dtype_in = qkv.dtype
+    if dtype_in == torch.float32:
+        qkv = qkv.to(torch.float16)  # need to truncate in case input is fp32
```

Comment:
Falcon is torch.bfloat16? Will the casting to float16 pose a problem.

Reply:
This is a minor modification of the old patching_code for llama which potentially can be trained in float32 (at least the small ones). The Old code always casted to float16 .. but flash_attention also supports bfloat16 .. I tested differences between always casting to float32 or as it is now keeping bfloat16 and the output logit-differences were minimal.

Reply:
Ah ok nice. I know float32 is used for the flash-attn softmax but the dtype_in should never be float32 unless you trained it with float32.

Reply:
correct, this conversion code is mainly for the case that models are trained with `dtype: float32` .. the patching_utils.py is used for all supported models. In the previous version the input was always converted with `.to(torch.float16)` .. and a potential reverse conversion happend in outer code (see here for [llama](https://github.com/LAION-AI/Open-Assistant/blob/64ece883f479ff185297cb8e9b33d2d8f78f2a2c/model/model_training/models/patching_llama.py#L62)) but I think it is better when the attention function returns in the same dtype that it got as input .. therefore I added the reverse dtype tracking and reverse conversion here.